community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -626,7 +623,17 @@ private void afterRollback()
     private void release()
     {
         locks.releaseAll();
-        pool.release( this );
+        if ( terminated )
+        {
+            // This transaction has been externally marked for termination.
+            // Just dispose of this transaction and don't return it to the pool.
+            dispose();
+        }
+        else
+        {
+            // Return this instance to the pool so that another transaction may use it.
+            pool.release( this );
+        }
```

Comment:
I'm not particularly against the given implementation, only that it muddles what it means to terminate a transaction. Perhaps you can expand the comment above the `dispose()` call to clarify that even though it is not strictly needed for cancelling queries, transaction termination is also used in mode switching, in which case it is very important that the transactions are not only cancelled, but that the pooled transaction objects are also disposed.